src/python_framework_v02/troute/nhd_network.py

```diff
@@ -172,11 +172,10 @@ def split_at_junction(network, path, node):
 
 
 def split_at_waterbodies_and_junctions(waterbody_nodes, network, path, node):
-    if path[-1] in waterbody_nodes:
-        return node in waterbody_nodes
+    if (path[-1] in waterbody_nodes) ^ (node in waterbody_nodes):
```

Comment:
bitwise XOR is sneaky here.  It works because the left and right evaluate to Booleans so the logical/bitwise result is the same.